core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/Collector.scala

```diff
@@ -170,5 +174,23 @@ trait Collector {
         unsecureEndpoint(routes)
         ()
     }
+
+    Runtime
+      .getRuntime
+      .addShutdownHook(new Thread(() => {
+        log.info("Received shutdown signal")
+        if (collectorConf.preTerminationUnhealthy) {
+          log.info("setting health endpoint to unhealthy")
+          health.toUnhealthy()
+        }
+        log.info(s"Sleeping for ${collectorConf.preTerminationPeriod}")
+        Thread.sleep(collectorConf.preTerminationPeriod.toMillis)
+        log.info("Initiating http server termination")
+        Await.result(system.terminate(), collectorConf.terminationDeadline)
+        log.info("Server terminated. Initiating sinks shutdown")
+        sinks.good.shutdown()
```

Comment:
What if some of hooks block or cannot complete at all - for example `sinkBatch` in `SqsSink` cannot succeed. Could there be a sensible timeout? Could we do `good` and `bad` sinks in parallel (although we do so much scary stuff with threads here already)?  Shouldn't it be part of `terminationDeadline`? What if `Await` fails to succeed - to me it seems that `good.shutdown` is a) more important; b) are we sure all sinks can flush and shutdown gracefully in absence of `system`? 

Another unrelated nit is that we likely want to log something of it with `warn` level - it's an important message and often people set logging severity to warning.

Reply:
Really good points.

My preference is to **not** have a timeout on the sinks.  Most orchestration platforms (e.g. kubernetes) will kill a container if it has not completed graceful shutdown within a configurable amount of time.  I think the orchestration platform is the proper place to configure this, rather than the application.

I do, however, agree it would be better to shut down the two sinks in parallel.  And I do agree it needs more error handling, so that sinks are shutdown even if actor system is not successfully terminated.  I will make those changes.

> are we sure all sinks can flush and shutdown gracefully in absence of system

Yes I think we are sure.  Because sinks are unaware that they run in the context of an akka system.